reconcile/ocm_machine_pools.py

```diff
@@ -1,137 +1,332 @@
 import json
 import logging
 import sys
+from abc import (
+    ABC,
+    abstractmethod,
+)
 from collections.abc import Mapping
-from typing import Any
+from typing import (
+    Iterable,
+    Optional,
+    Tuple,
+)
+
+from pydantic import BaseModel
 
 from reconcile import queries
+from reconcile.gql_definitions.common.clusters import (
+    ClusterMachinePoolV1,
+    ClusterV1,
+)
+from reconcile.gql_definitions.common.clusters import query as clusters_query
+from reconcile.utils import gql
 from reconcile.utils.disabled_integrations import integration_is_enabled
-from reconcile.utils.ocm import OCMMap
+from reconcile.utils.ocm import (
+    OCM,
+    OCMMap,
+)
 
 QONTRACT_INTEGRATION = "ocm-machine-pools"
 
 
-def fetch_current_state(clusters):
-    settings = queries.get_app_interface_settings()
-    ocm_map = OCMMap(
-        clusters=clusters, integration=QONTRACT_INTEGRATION, settings=settings
-    )
+class InvalidUpdateError(Exception):
+    pass
+
+
+class AbstractPool(ABC, BaseModel):
+    id: str
+    replicas: int
+    taints: Optional[list[Mapping[str, str]]]
+    labels: Optional[Mapping[str, str]]
+    cluster: str
+
+    @abstractmethod
+    def create(self, ocm: OCM) -> None:
+        pass
+
+    @abstractmethod
+    def delete(self, ocm: OCM) -> None:
+        pass
+
+    @abstractmethod
+    def update(self, ocm: OCM) -> None:
+        pass
+
+    @abstractmethod
+    def has_diff(self, pool: ClusterMachinePoolV1) -> bool:
+        pass
+
+    @abstractmethod
+    def invalid_diff(self, pool: ClusterMachinePoolV1) -> Optional[str]:
+        pass
+
+
+class MachinePool(AbstractPool):
+    instance_type: str
+
+    def delete(self, ocm: OCM) -> None:
+        ocm.delete_machine_pool(self.cluster, self.dict(by_alias=True))
+
+    def create(self, ocm: OCM) -> None:
+        ocm.create_machine_pool(self.cluster, self.dict(by_alias=True))
+
+    def update(self, ocm: OCM) -> None:
+        ocm.update_machine_pool(self.cluster, self.dict(by_alias=True))
+
+    def has_diff(self, pool: ClusterMachinePoolV1) -> bool:
+        return (
+            self.replicas != pool.replicas
+            or self.taints != pool.taints
+            or self.labels != pool.labels
```

Comment:
should other field be taken into account, like machine type? or is that forbidden ?

Reply:
ah, i should have read just below i suppose :)